backend/server/routers/auth.py

```diff
@@ -1,14 +1,276 @@
+            detail="Could not exchange tokens, contact admin please.",
+        ) from e
+
+    # insert new user into database if it does not exist, will collide if it already does which is good
+    uid = id_token["sub"]
+    insert_new_user(uid, NotSetupUserStorage(guest=False))
+
+    # create new login session for user in db, generating new tokens
+    new_oidc_info = SessionOIDCInfoModel(
+        access_token=tokens["access_token"],
+        raw_id_token=tokens["id_token"],
+        refresh_token=tokens["refresh_token"],
+        validated_id_token=cast(dict, id_token),
+    )
+    new_session_token, session_expiry, new_refresh_token, refresh_expiry = setup_new_csesoc_session(uid, new_oidc_info)
+
+    # set the cookies and return the identity
+    set_secure_cookie(res, AUTH_STATE_COOKIE, None)
+    set_secure_cookie(res, REFRESH_TOKEN_COOKIE, new_refresh_token, refresh_expiry)
+    return IdentityPayload(session_token=new_session_token, exp=session_expiry, uid=uid)
+
+
+
+@router.post("/logout", response_model=None)
+def logout(res: Response, token: Annotated[SessionToken, Security(require_token)], refresh_token: Annotated[Optional[RefreshToken], Cookie(alias=REFRESH_TOKEN_COOKIE)] = None):
+    # delete the cookie first since this should always happen
+    set_secure_cookie(res, REFRESH_TOKEN_COOKIE, None)
+
+    session = _try_get_session_info_for_logout(token, refresh_token)
+    if session is None:
+        raise HTTPException(
+            status_code=HTTP_401_UNAUTHORIZED,
+            detail="Provided token was expired, please re-authenticate.",
+            headers={ "WWW-Authenticate": "Bearer", "set-cookie": res.headers["set-cookie"] },
+        )
+
+    # kill our session, then revoke the oidc tokens
+    # done in this order since once our session is destroyed, their oidc tokens are gone anyway
+    sid, session_info = session
+    assert logout_session(sid)
+
+    if session_info.type == "guest":
+        # a guest user only ever gets one login session, so we can safely drop their data after a successful logout
+        assert delete_user(session_info.uid)
+    if session_info.type == "csesoc":
+        # only need to revoke a token for fed auth sessions
+        try:
+            revoke_token(session_info.oidc_info.refresh_token, "refresh_token")
+        except OIDCInvalidGrant as e:
+            # invalid grant could happen if the oidc token is expired, thats fine
+            # NOTE: i dont think we want to throw error if it is invalid? but its fine i guess
+            raise HTTPException(
+                status_code=HTTP_401_UNAUTHORIZED,
+                detail="Your csesoc session has expired, please re-login.",
+                headers={ "WWW-Authenticate": "Bearer", "set-cookie": res.headers["set-cookie"] },
+            ) from e
+        except OIDCTokenError as e:
+            # cant imagine what error this could be, but if it happens, its bad
+            print(e)
```

Comment:
Do we currently have a way to track any internal server errors that take place in prod?

Reply:
i dont think so, a proper logger instance and usage is definitely needed (and something I have on the list of post merge todos)